src/common/ad_mem.c

```diff
@@ -3,5 +3,1857 @@
+	for (i = 0; i < bd->bd_asp_nr; i++)
+		blob->bb_arena_last[i] = bd->bd_asp[i].as_arena_last;
+out:
+	return rc;
+}
+
+static uint64_t
+blob_incarnation(struct ad_blob *blob)
+{
+	return blob->bb_df->bd_incarnation;
+}
+
+static void
+blob_set_opened(struct ad_blob *blob)
+{
+	struct ad_blob_df *bd = blob->bb_df;
+
+	bd->bd_incarnation = d_timeus_secdiff(0);
+	bd->bd_back_ptr	   = (unsigned long)blob;
+
+	blob->bb_opened = 1;
+	if (blob->bb_dummy) {
+		D_ASSERT(!dummy_blob);
+		dummy_blob = blob;
+	}
+}
+
+/**
+ * API design is a little bit confusing for now, the process is like:
+ * - VOS calls umem_pool_create()->blob_prep_create()
+ * - VOS initialize BIO context
+ * - VOS calls blob_post_create() to read blob (needs BIO context)
+ */
+int
+ad_blob_prep_create(char *path, daos_size_t size, struct ad_blob_handle *bh)
+{
+	struct ad_blob	*blob;
+	bool		 is_dummy = false;
+
+	if (!strcmp(path, DUMMY_BLOB)) {
+		if (dummy_blob)
+			return -DER_EXIST;
+		is_dummy = true;
+	}
+	D_ASSERT(is_dummy); /* XXX the only thing can be supported now */
+
+	D_ALLOC_PTR(blob);
+	if (!blob)
+		return -DER_NOMEM;
+
+	blob->bb_store.stor_size = size;
+	blob->bb_fd  = -1; /* XXX, create file, falloc */
+	blob->bb_ref = 1;
+	blob->bb_dummy = is_dummy;
+
+	bh->bh_blob = blob;
+	return 0;
+}
+
+int
+ad_blob_post_create(struct ad_blob_handle bh)
+{
+	struct ad_blob		*blob = bh.bh_blob;
+	struct umem_store	*store = &blob->bb_store;
+	struct ad_blob_df	*bd;
+	struct ad_arena		*arena;
+	struct umem_store_iod	 iod;
+	d_sg_list_t		 sgl;
+	d_iov_t			 iov;
+	int			 rc;
+
+	/* NB: store::stor_size can be changed by upper level stack */
+	blob->bb_pgs_nr = ((blob_size(blob) + ARENA_SIZE_MASK) >> ARENA_SIZE_BITS);
+
+	rc = blob_init(blob);
+	if (rc)
+		return rc;
+
+	bd = blob->bb_df;
+	bd->bd_magic		= BLOB_MAGIC;
+	bd->bd_version		= AD_MEM_VERSION;
+	bd->bd_addr		= store->stor_addr;
+	bd->bd_size		= blob_size(blob);
+	bd->bd_arena_size	= ARENA_SIZE;
+
+	/* register the default arena */
+	rc = blob_register_arena(blob, 0, grp_specs_def, ARRAY_SIZE(grp_specs_def), NULL);
+	D_ASSERT(rc == 0); /* no reason to fail */
+
+	/* create arena 0 (ad_blob_df is stored in the first 32K of it) */
+	rc = arena_reserve(blob, 0, NULL, &arena);
+	D_ASSERT(rc == 0);
+	D_ASSERT(arena->ar_df);
+	D_ASSERT(arena->ar_df->ad_id == 0);
+
+	/* NB: no transaction, write arena[0] and super block to storage straight away */
+	rc = arena_tx_publish(arena, NULL);
+	if (rc)
+		goto failed;
+
+	arena->ar_unpub = false;
+
+	blob->bb_arena_last[0]	= bd->bd_asp[0].as_arena_last;
+	/* already published arena[0], clear the reserved bit */
+	clrbit(blob->bb_bmap_rsv, arena->ar_df->ad_id);
+
+	ad_iod_set(&iod, blob_ptr2addr(blob, arena->ar_df), ARENA_HDR_SIZE + BLOB_HDR_SIZE);
+	ad_sgl_set(&sgl, &iov, arena->ar_df, ARENA_HDR_SIZE + BLOB_HDR_SIZE);
+
+	D_ASSERT(store->stor_ops);
+	rc = store->stor_ops->so_write(store, &iod, &sgl);
```

Comment:
I'm wondering what will be stored in such a large superblock, since no WAL to guarantee the atomicity of this 32k superblock update, you need to update the 4k block contains magic after all other block updates done, so that we could tell the superblock consistent by check the magic on load time.

And I'm not sure why we don't rely on WAL to write ad superblock? because it's only updated on format time?

Reply:
This is the format, so let's do it in a simple way, after the format, all the update to superblock will go through WAL.